server/db/rethinkdb/adapter.go

```diff
@@ -1518,17 +1524,69 @@ func (a *adapter) SubsUpdate(topic string, user t.Uid, update map[string]interfa
 // SubsDelete marks subscription as deleted.
 func (a *adapter) SubsDelete(topic string, user t.Uid) error {
 	now := t.TimeNow()
+	forUser := user.String()
+
+	// Mark subscription as deleted.
 	_, err := rdb.DB(a.dbName).Table("subscriptions").
-		Get(topic + ":" + user.String()).Update(map[string]interface{}{
+		Get(topic + ":" + forUser).Update(map[string]interface{}{
 		"UpdatedAt": now,
 		"DeletedAt": now,
 	}).RunWrite(a.conn)
-	// _, err := rdb.DB(a.dbName).Table("subscriptions").Get(topic + ":" + user.String()).Delete().RunWrite(a.conn)
-	return err
+
+	if err != nil {
+		return err
+	}
+
+	if t.IsChannel(topic) {
+		// Channel readers cannot delete messages, all done.
+		log.Println("this is a channel, all done")
+		return nil
+	}
+
+	// Remove records of deleted messages.
+
+	// Delete dellog entries of the current user.
+	resp, err := rdb.DB(a.dbName).Table("dellog").
+		// Select all log entries for the given table.
+		Between([]interface{}{topic, rdb.MinVal}, []interface{}{topic, rdb.MaxVal},
+			rdb.BetweenOpts{Index: "Topic_DelId"}).
+		// Keep entries soft-deleted for the current user only.
+		Filter(rdb.Row.Field("DeletedFor").Eq(forUser)).
+		// Delete them.
+		Delete().
+		RunWrite(a.conn)
+
+	if err != nil || resp.Deleted == 0 {
+		// Either an error or nothing was deleted. Not much we can do with the error.
+		// Returning nil even on failure.
+		return nil
+	}
+
+	// Remove current user from the messages' soft-deletion lists.
+	_, err = rdb.DB(a.dbName).Table("messages").
+		// Select all messages in the given topic.
+		Between(
+			[]interface{}{topic, forUser, rdb.MinVal},
+			[]interface{}{topic, forUser, rdb.MaxVal},
+			rdb.BetweenOpts{Index: "Topic_DeletedFor"}).
+		// Update the field DeletedFor:
+		Update(map[string]interface{}{
+			// Take the DeletedFor array, subtract all values which contain current user ID in 'User' field.
+			"DeletedFor": rdb.Row.Field("DeletedFor").
+				SetDifference(
+					rdb.Row.Field("DeletedFor").
+						Filter(map[string]interface{}{"User": forUser}))}).
+		RunWrite(a.conn)
+
+	if err != nil {
+		log.Println("error 2", err)
```

Comment:
Return an error here?

Reply:
Sorry. I was pretty sure I cleaned it up but apparently forgot to push.